File: minesweeper.java
// minesweeper.java
import java.io.*;
import java.util.*;

public class minesweeper {
    private static final String RESET = "\u001B[0m";
    private static final String RED = "\u001B[91m";
    private static final String GREEN = "\u001B[92m";
    private static final String YELLOW = "\u001B[93m";
    private static final String BLUE = "\u001B[94m";
    private static final String MAGENTA = "\u001B[95m";
    private static final String CYAN = "\u001B[96m";
    private static final String WHITE = "\u001B[97m";
    private static final String GRAY = "\u001B[90m";
    private static final String BOLD = "\u001B[1m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    static class MinesweeperGame {
        int rows, cols, minesTotal, minesRemaining;
        int[][] real;
        boolean[][] revealed, flags;
        boolean gameOver, won, firstMove;
        int moves;
        long startTime;

        MinesweeperGame(int rows, int cols, int mines) {
            this.rows = rows;
            this.cols = cols;
            minesTotal = mines;
            minesRemaining = mines;
            real = new int[rows][cols];
            revealed = new boolean[rows][cols];
            flags = new boolean[rows][cols];
            firstMove = true;
        }

        void generateBoard(int firstRow, int firstCol) {
            Random rand = new Random();
            int placed = 0;
            while (placed < minesTotal) {
                int r = rand.nextInt(rows);
                int c = rand.nextInt(cols);
                if (r == firstRow && c == firstCol) continue;
                if (Math.abs(r - firstRow) <= 1 && Math.abs(c - firstCol) <= 1) continue;
                if (real[r][c] == -1) continue;
                real[r][c] = -1;
                placed++;
            }
            for (int r = 0; r < rows; r++) {
                for (int c = 0; c < cols; c++) {
                    if (real[r][c] == -1) continue;
                    int cnt = 0;
                    for (int dr = -1; dr <= 1; dr++) {
                        for (int dc = -1; dc <= 1; dc++) {
                            int nr = r + dr, nc = c + dc;
                            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && real[nr][nc] == -1)
                                cnt++;
                        }
                    }
                    real[r][c] = cnt;
                }
            }
        }

        void render() {
            System.out.print("\033[H\033[2J");
            System.out.flush();
            System.out.print(colorize("   ", BOLD));
            for (int c = 0; c < cols; c++) {
                System.out.print(colorize(String.format("%2d ", c+1), BOLD));
            }
            System.out.println();
            System.out.print(colorize("   +", GRAY));
            for (int c = 0; c < cols; c++) System.out.print("---");
            System.out.println(colorize("+", GRAY));
            for (int r = 0; r < rows; r++) {
                System.out.print(colorize(String.format("%2d ", r+1), BOLD));
                System.out.print(colorize("|", GRAY));
                for (int c = 0; c < cols; c++) {
                    if (gameOver && real[r][c] == -1) {
                        System.out.print(colorize(" 💣", RED));
                    } else if (flags[r][c]) {
                        System.out.print(colorize(" ⚑", YELLOW));
                    } else if (revealed[r][c]) {
                        int val = real[r][c];
                        if (val == -1) System.out.print(colorize(" 💣", RED));
                        else if (val == 0) System.out.print("  ");
                        else {
                            String[] colors = {"", BLUE, GREEN, RED, MAGENTA, CYAN, YELLOW, WHITE, GRAY};
                            System.out.print(colorize(" " + val, colors[val]));
                        }
                    } else {
                        System.out.print(" ■");
                    }
                }
                System.out.println(colorize("|", GRAY));
            }
            System.out.print(colorize("   +", GRAY));
            for (int c = 0; c < cols; c++) System.out.print("---");
            System.out.println(colorize("+", GRAY));
            System.out.println(colorize("Мин осталось: " + minesRemaining, YELLOW));
            if (!gameOver && startTime != 0) {
                long elapsed = (System.currentTimeMillis() - startTime) / 1000;
                System.out.println(colorize("Время: " + elapsed + " сек", BLUE));
            }
            if (gameOver) {
                System.out.println(won ? colorize("🎉 ПОБЕДА!", GREEN) : colorize("💥 ПОРАЖЕНИЕ!", RED));
            }
        }

        void openCell(int r, int c) {
            if (gameOver || revealed[r][c] || flags[r][c]) return;
            if (firstMove) {
                generateBoard(r, c);
                firstMove = false;
                startTime = System.currentTimeMillis();
            }
            revealed[r][c] = true;
            moves++;
            if (real[r][c] == -1) {
                gameOver = true;
                return;
            }
            if (real[r][c] == 0) {
                for (int dr = -1; dr <= 1; dr++) {
                    for (int dc = -1; dc <= 1; dc++) {
                        int nr = r + dr, nc = c + dc;
                        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && !revealed[nr][nc] && !flags[nr][nc]) {
                            openCell(nr, nc);
                        }
                    }
                }
            }
            int revealedCount = 0;
            for (int i = 0; i < rows; i++)
                for (int j = 0; j < cols; j++)
                    if (revealed[i][j]) revealedCount++;
            if (revealedCount == rows * cols - minesTotal) {
                won = true;
                gameOver = true;
            }
        }

        void toggleFlag(int r, int c) {
            if (gameOver || revealed[r][c]) return;
            if (flags[r][c]) {
                flags[r][c] = false;
                minesRemaining++;
            } else {
                flags[r][c] = true;
                minesRemaining--;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        int rows = 9, cols = 9, mines = 10;
        if (args.length > 0) {
            String level = args[0].toLowerCase();
            if (level.equals("easy")) { rows = 9; cols = 9; mines = 10; }
            else if (level.equals("medium")) { rows = 16; cols = 16; mines = 40; }
            else if (level.equals("hard")) { rows = 30; cols = 16; mines = 99; }
            else {
                try {
                    int v = Integer.parseInt(level);
                    rows = cols = v;
                    if (rows < 3) rows = cols = 3;
                    if (args.length > 1) mines = Integer.parseInt(args[1]);
                    else mines = Math.max(1, rows * cols / 10);
                } catch (NumberFormatException e) {
                    System.out.println("Неверный аргумент.");
                    return;
                }
            }
        }
        MinesweeperGame game = new MinesweeperGame(rows, cols, mines);
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        while (!game.gameOver) {
            game.render();
            System.out.print("Введите команду (строка столбец для открытия, f строка столбец для флага, q для выхода): ");
            String cmd = reader.readLine().trim();
            if (cmd.equals("q")) {
                System.out.println("Выход.");
                return;
            }
            String[] parts = cmd.split("\\s+");
            if (parts.length < 2) {
                System.out.println("Неверный формат.");
                continue;
            }
            if (parts[0].equals("f")) {
                if (parts.length != 3) {
                    System.out.println("Неверный формат. Используйте: f строка столбец");
                    continue;
                }
                try {
                    int r = Integer.parseInt(parts[1]) - 1;
                    int c = Integer.parseInt(parts[2]) - 1;
                    if (r >= 0 && r < rows && c >= 0 && c < cols)
                        game.toggleFlag(r, c);
                    else
                        System.out.println("Координаты вне поля.");
                } catch (NumberFormatException e) {
                    System.out.println("Неверный формат.");
                }
            } else {
                if (parts.length != 2) {
                    System.out.println("Неверный формат. Используйте: строка столбец");
                    continue;
                }
                try {
                    int r = Integer.parseInt(parts[0]) - 1;
                    int c = Integer.parseInt(parts[1]) - 1;
                    if (r >= 0 && r < rows && c >= 0 && c < cols)
                        game.openCell(r, c);
                    else
                        System.out.println("Координаты вне поля.");
                } catch (NumberFormatException e) {
                    System.out.println("Неверный формат.");
                }
            }
        }
        game.render();
        if (game.won) {
            long elapsed = (System.currentTimeMillis() - game.startTime) / 1000;
            System.out.printf("Поздравляем! Вы выиграли за %d ходов и %d секунд.\n", game.moves, elapsed);
        } else {
            System.out.println("Игра окончена. Попробуйте снова!");
        }
    }
}

File: minesweeper.go
// minesweeper.go
package main

import (
	"bufio"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

const (
	reset  = "\033[0m"
	red    = "\033[91m"
	green  = "\033[92m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	magenta= "\033[95m"
	cyan   = "\033[96m"
	white  = "\033[97m"
	gray   = "\033[90m"
	bold   = "\033[1m"
)

func colorize(text, color string) string {
	return color + text + reset
}

type Minesweeper struct {
	rows, cols, minesTotal int
	board                   [][]string   // видимое поле
	real                    [][]int      // -1 = мина, 0-8 = число
	revealed                [][]bool
	flags                   [][]bool
	gameOver                bool
	won                     bool
	firstMove               bool
	minesRemaining          int
	startTime               time.Time
	moves                   int
}

func NewMinesweeper(rows, cols, mines int) *Minesweeper {
	m := &Minesweeper{
		rows:           rows,
		cols:           cols,
		minesTotal:     mines,
		minesRemaining: mines,
		firstMove:      true,
	}
	m.board = make([][]string, rows)
	m.real = make([][]int, rows)
	m.revealed = make([][]bool, rows)
	m.flags = make([][]bool, rows)
	for i := 0; i < rows; i++ {
		m.board[i] = make([]string, cols)
		m.real[i] = make([]int, cols)
		m.revealed[i] = make([]bool, cols)
		m.flags[i] = make([]bool, cols)
	}
	return m
}

func (m *Minesweeper) generateBoard(firstRow, firstCol int) {
	mines := make(map[[2]int]bool)
	for len(mines) < m.minesTotal {
		r := rand.Intn(m.rows)
		c := rand.Intn(m.cols)
		if r == firstRow && c == firstCol {
			continue
		}
		if abs(r-firstRow) <= 1 && abs(c-firstCol) <= 1 {
			continue
		}
		key := [2]int{r, c}
		if !mines[key] {
			mines[key] = true
			m.real[r][c] = -1
		}
	}
	// Подсчёт чисел
	for r := 0; r < m.rows; r++ {
		for c := 0; c < m.cols; c++ {
			if m.real[r][c] == -1 {
				continue
			}
			count := 0
			for dr := -1; dr <= 1; dr++ {
				for dc := -1; dc <= 1; dc++ {
					nr, nc := r+dr, c+dc
					if nr >= 0 && nr < m.rows && nc >= 0 && nc < m.cols && m.real[nr][nc] == -1 {
						count++
					}
				}
			}
			m.real[r][c] = count
		}
	}
}

func (m *Minesweeper) render() {
	fmt.Print("\033[H\033[2J")
	fmt.Print(colorize("   ", bold))
	for c := 0; c < m.cols; c++ {
		fmt.Print(colorize(fmt.Sprintf("%2d ", c+1), bold))
	}
	fmt.Println()
	fmt.Print(colorize("   "+"+", gray))
	for c := 0; c < m.cols; c++ {
		fmt.Print("---")
	}
	fmt.Println(colorize("+", gray))
	for r := 0; r < m.rows; r++ {
		fmt.Print(colorize(fmt.Sprintf("%2d ", r+1), bold))
		fmt.Print(colorize("|", gray))
		for c := 0; c < m.cols; c++ {
			if m.gameOver && m.real[r][c] == -1 {
				fmt.Print(colorize(" 💣", red))
			} else if m.flags[r][c] {
				fmt.Print(colorize(" ⚑", yellow))
			} else if m.revealed[r][c] {
				val := m.real[r][c]
				if val == -1 {
					fmt.Print(colorize(" 💣", red))
				} else if val == 0 {
					fmt.Print("  ")
				} else {
					colors := []string{"", blue, green, red, magenta, cyan, yellow, white, gray}
					fmt.Print(colorize(fmt.Sprintf(" %d", val), colors[val]))
				}
			} else {
				fmt.Print(" ■")
			}
		}
		fmt.Println(colorize("|", gray))
	}
	fmt.Print(colorize("   "+"+", gray))
	for c := 0; c < m.cols; c++ {
		fmt.Print("---")
	}
	fmt.Println(colorize("+", gray))
	fmt.Println(colorize(fmt.Sprintf("Мин осталось: %d", m.minesRemaining), yellow))
	if !m.gameOver && !m.startTime.IsZero() {
		elapsed := int(time.Since(m.startTime).Seconds())
		fmt.Println(colorize(fmt.Sprintf("Время: %d сек", elapsed), blue))
	}
	if m.gameOver {
		if m.won {
			fmt.Println(colorize("🎉 ПОБЕДА!", green))
		} else {
			fmt.Println(colorize("💥 ПОРАЖЕНИЕ!", red))
		}
	}
}

func (m *Minesweeper) openCell(r, c int) {
	if m.gameOver || m.revealed[r][c] || m.flags[r][c] {
		return
	}
	if m.firstMove {
		m.generateBoard(r, c)
		m.firstMove = false
		m.startTime = time.Now()
	}
	m.revealed[r][c] = true
	m.moves++

	if m.real[r][c] == -1 {
		m.gameOver = true
		return
	}
	if m.real[r][c] == 0 {
		for dr := -1; dr <= 1; dr++ {
			for dc := -1; dc <= 1; dc++ {
				nr, nc := r+dr, c+dc
				if nr >= 0 && nr < m.rows && nc >= 0 && nc < m.cols && !m.revealed[nr][nc] && !m.flags[nr][nc] {
					m.openCell(nr, nc)
				}
			}
		}
	}
	// Проверка победы
	revealedCount := 0
	for i := 0; i < m.rows; i++ {
		for j := 0; j < m.cols; j++ {
			if m.revealed[i][j] {
				revealedCount++
			}
		}
	}
	if revealedCount == m.rows*m.cols-m.minesTotal {
		m.won = true
		m.gameOver = true
	}
}

func (m *Minesweeper) toggleFlag(r, c int) {
	if m.gameOver || m.revealed[r][c] {
		return
	}
	if m.flags[r][c] {
		m.flags[r][c] = false
		m.minesRemaining++
	} else {
		m.flags[r][c] = true
		m.minesRemaining--
	}
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func main() {
	rand.Seed(time.Now().UnixNano())
	var rows, cols, mines int = 9, 9, 10
	if len(os.Args) > 1 {
		arg := os.Args[1]
		switch arg {
		case "easy":
			rows, cols, mines = 9, 9, 10
		case "medium":
			rows, cols, mines = 16, 16, 40
		case "hard":
			rows, cols, mines = 30, 16, 99
		default:
			if v, err := strconv.Atoi(arg); err == nil {
				rows, cols = v, v
				if rows < 3 {
					rows, cols = 3, 3
				}
				if len(os.Args) > 2 {
					if m, err := strconv.Atoi(os.Args[2]); err == nil {
						mines = m
					}
				} else {
					mines = max(1, rows*cols/10)
				}
			}
		}
	}
	if len(os.Args) > 2 && os.Args[1] != "easy" && os.Args[1] != "medium" && os.Args[1] != "hard" {
		if m, err := strconv.Atoi(os.Args[2]); err == nil {
			mines = m
		}
	}

	game := NewMinesweeper(rows, cols, mines)
	scanner := bufio.NewScanner(os.Stdin)

	for !game.gameOver {
		game.render()
		fmt.Print("Введите команду (строка столбец для открытия, f строка столбец для флага, q для выхода): ")
		if !scanner.Scan() {
			break
		}
		cmd := strings.TrimSpace(scanner.Text())
		if cmd == "q" {
			fmt.Println("Выход.")
			return
		}
		parts := strings.Fields(cmd)
		if len(parts) < 2 {
			fmt.Println("Неверный формат.")
			continue
		}
		if parts[0] == "f" {
			if len(parts) != 3 {
				fmt.Println("Неверный формат. Используйте: f строка столбец")
				continue
			}
			r, err1 := strconv.Atoi(parts[1])
			c, err2 := strconv.Atoi(parts[2])
			if err1 == nil && err2 == nil && r >= 1 && r <= rows && c >= 1 && c <= cols {
				game.toggleFlag(r-1, c-1)
			} else {
				fmt.Println("Координаты вне поля.")
			}
		} else {
			if len(parts) != 2 {
				fmt.Println("Неверный формат. Используйте: строка столбец")
				continue
			}
			r, err1 := strconv.Atoi(parts[0])
			c, err2 := strconv.Atoi(parts[1])
			if err1 == nil && err2 == nil && r >= 1 && r <= rows && c >= 1 && c <= cols {
				game.openCell(r-1, c-1)
			} else {
				fmt.Println("Координаты вне поля.")
			}
		}
	}
	game.render()
	if game.won {
		elapsed := int(time.Since(game.startTime).Seconds())
		fmt.Printf("Поздравляем! Вы выиграли за %d ходов и %d секунд.\n", game.moves, elapsed)
	} else {
		fmt.Println("Игра окончена. Попробуйте снова!")
	}
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
